### ISSUE TITLE
Invalid `inject()` function typing

### ISSUE THREAD

**Issue by yjaaidi** [angular/angular#46251]
### Which @angular/* package(s) are the source of the bug?

core

### Is this a regression?

No

### Description

`inject()`'s type description seems to be invalid:
https://github.com/angular/angular/blob/5cfde8b23e72cff4c9359f765a462b474bb32631/packages/core/src/di/injector_compatibility.ts#L78-L80

Return type should be `T | null` only if flag is set to `InjectFlags.Optional` meaning that the signature should be something like this:

```typescript
 export function ɵɵinject<T>(token: ProviderToken<T>): T; 
 export function ɵɵinject<T>(token: ProviderToken<T>, flags?: InjectFlags.Default | InjectFlags...): T; 
 export function ɵɵinject<T>(token: ProviderToken<T>, flags?: InjectFlags.Optional): T | null; 
 export function ɵɵinject<T>(token: ProviderToken<T>, flags = InjectFlags.Default): T|null
```

**Comment by JoostK** [comment]
This would be a feature request for a narrower return type of `inject`, as the current signatures are not invalid. They just don't make a distinction between various injection flags.

Since `InjectFlags` is a bit mask, this requires special care. The following set of signature overloads makes this work as desired:

```ts
declare function inject<T>(token: Type<T>): T;
declare function inject<T>(token: Type<T>, flags: InjectFlags.Optional): T|null;
declare function inject<T>(token: Type<T>, flags: InjectFlags.Default | InjectFlags.Host | InjectFlags.Self | InjectFlags.SkipSelf): T;
declare function inject<T>(token: Type<T>, flags: InjectFlags): T|null;
```

[TypeScript playground](https://www.typescriptlang.org/play?ts=4.7.2#code/KYOwrgtgBAkiBWwDGAXAYgGwIYHMDOUA3gFBRQAiwAZlmBilALxQAMANKVABID2eDzAIwcyAZWAYqTKACYRUUQGsAlgAdxk6QBZ5AeVUplPEFgzSAHBwC+xYsAAeqngCcGykCmDOaSYFAAqAJ6qwAA8-gB8UA6eIAAmBGhgIKhGIEScIMAA7gAUAHSFWM74AFxQWCCBANoAugCU5f4A3MQ2xHHI2M5+VMmpxlDuiKjhEbkoPIqgTcFhkY0BrZ1I3b39hoPDyChjE1MzAXNjbFBU2GWwCDuYuHj5+psmGIv+AD7gGBjLXcXrKU8htdRpF9tMQLMQiczhc8OU4CN0LD8pQaHQGG8rojbvh8rx+FBMQibsiNFIicCkXd8ko1GTXj9Vn8zhs0kDEXtJuDIfMIqdznd4ZScXhXh86N9bEy8ARxM4AG7KXxEdpIYwEvAqdQSKTMbaoXJyxW+U7E1Aimla+mtAD0NrIZAAegB+KXqhg8AxpUzSfUoQ1eY3AU3C5GPb0vW32h0ut0gAmdNH0ABqvspAYVSuDWJJ1NRtHo9SjDqgseIavjDAAFnwUABBeLh4w+vXpo1ZkPY5H4jE581hr3NyPEO0lsujgByPE85QrvgMYFMGECUBQVeUBDVEAARu4sICeFJdygBfgoHEeMA8CAAOQMCBYaZQPCgF+nZQoG8EdyrMCdOJQFWXh+JMz7AH42RVvugFYKoITxmcLgVFAx4PpqUDZB+NZgAwTbPOW7rPlaOoNnEZJpoiGZBp2ua4rS2qaBSXbUtaI7Rk6rodL8PRQBWBKnnCfZUvgrR8W48TKD0-Z3BROxUR2MJ3EWbFjq6QA)

What is concerning here though is that TypeScript always selects the _second_ overload (the one with `InjectFlags.Optional`) for inline bit masks, as can be seen in the `skipSelfAndSelf` case in the playground. This is unlike using a variable of type `InjectFlags`, which does resolve to the _last_ overload as desired.

Currently this sort of works out, especially in practice as it doesn't make sense to combine any of the `Self`, `SkipSelf` and `Host` flags into a single call. It does concern me though that TypeScript doesn't select the last overload for `skipSelfAndSelf`, as it indicates that TypeScript's overload resolution isn't equipped for handling this properly at the moment. It also means that the specific ordering of overload 2 and 3 is important, which ideally it should not be (the signatures don't have overlap in their `flags` parameter, so their relative ordering shouldn't affect overload resolution but it does).

**Comment by yjaaidi** [comment]
Thanks @JoostK for diving so deeply into this & also taking care of the repro 😅.
I've been playing with it and the TypeScript resolution behavior is so intriguing.
It simply seems that bitwise operators resolve to `number` which is compatible with any `InjectFlags` value, that's why it picks the first signature with a flag parameter:
<img width="393" alt="image" src="https://user-images.githubusercontent.com/2674658/172010685-f622ea09-960d-4775-a8cb-70b1d9435429.png">

Meanwhile, this gets fixed in TypeScript, I think that the best solution is the one you came out with.
It reduces the use cases where users will have to cast the return type to flag combinations.

Do you want me to go ahead with a PR?

Thanks for your time

**Comment by JoostK** [comment]
A PR is welcome, that would allow for a TGP (running all g3 tests) to see what the impact of such a change would be.

**Comment by yjaaidi** [comment]
I'm taking care of it.

**Comment by alxhub** [comment]
I think the right answer here is deprecate the current `inject` format with the bitfield options and add a new one which uses an `options` object instead.

**Comment by dario-piotrowicz** [comment]
@yjaaidi I think that @alxhub is quite an interesting option, I have been chatting with @AndrewKushnir who also stands by this idea proposing that we could indeed convert the current:
```ts
inject(MyDep, InjectFlags.Optional & InjectFlags.Self)
// or
injector.get(MyDep, InjectFlags.Optional & InjectFlags.Self)
```
into something like:
```ts
inject(MyDep, {optional: true, self: true})
// or
injector.get(MyDep, {optional: true, self: true})
```

We should probably create an issue with such a proposal (which we could then work on when there is a consensus) and think of the different types of implementations (objects, tuples, a new enum, etc..) and strategies (for example, if using objects, can we also add some validation so that conflicting flags can't be set at the same time?)

@yjaaidi what do you think? would you like to create such issue? :slightly_smiling_face:
(if not I would be happy to help with that)

**Comment by yjaaidi** [comment]
> I think the right answer here is deprecate the current inject format with the bitfield options and add a new one which uses an options object instead.

I can't agree more with @alxhub . I'm not a big fan of bitfield options but I thought that it was there for some perf optimization. If you don't need it, I'll be the last person to defend it 😁

**Comment by yjaaidi** [comment]
Hey @dario-piotrowicz! Yes totally, but we can probably just work on this issue instead of opening a new one. What do you think?

**Comment by yjaaidi** [comment]
Are these the right combinations?

```typescript
type InjectOptions = ({} | {self: true} | {skipSelf: true} | {host: true, skipSelf?: true})
type Optional = {optional: true}

function inject<T>(token: Type<T>): T;
function inject<T>(token: Type<T>, options: InjectOptions & Optional): T | null;
function inject<T>(token: Type<T>, options: InjectOptions): T;
function inject<T>(token: Type<T>, flags: InjectFlags): T | null;
function inject<T>(token: Type<T>, flagsOrOptions?: unknown) {
}

class X {}

const a: X  = inject(X);
const b: X  = inject(X, {self: true});
const c: X  = inject(X, {skipSelf: true});
const d: X  = inject(X, {host: true});
const e: X  = inject(X, {host: true, skipSelf: true});
const f: X | null  = inject(X, {optional: true});
const g: X | null  = inject(X, {self: true, optional: true});
const h: X | null  = inject(X, {skipSelf: true, optional: true});
const i: X | null  = inject(X, {host: true, optional: true});
const j: X | null  = inject(X, {host: true, skipSelf: true, optional: true});

```


https://www.typescriptlang.org/play?ts=4.7.2#code/FASwdgLgpgTgZgQwMZQAQBUCeAHKAedAPlSgA9owATAZ1QDEBXMJCEAezFQG9hVUwoAd1QAKAHQSEMAObUAXKgRhMAbQC6ASgXoA3MAC+wCDjQBJMACsoLOgBsEs1AF5+DALYAjWHuBGTqcysWAHlsVg5aFxEufVQAH25qKFs4BQgYBihYhK5qAGsQbABlZNTUdMzs7gALNmoINIyoABpUfMKSlIB+RsqNP1xUUPCwBFtnbjYw9lHbXqzfOCYWGdRwIIgCQhEINjyoMG0TLa0MPSXmEbXLa02iHb2Do9wt1qmR+QCbkOmI1AAyIa-WandDxVy2WznZZXda3LYPfaHDDHIhvYGfQK3YYzaig6GXVZwlgI3ZI574NGoOD2WQKLE2Wl47TgsAMSEElYca4bUmPZFYF5UmkOajBGA4iI9VBMPJgNiCMAabgGXxIezUWgADW4hmASAiEEUCh1E2JEBEWo0egNYHqqA8Jr4LnNltauVK8301v1htQSCdZu+Fq17vaxU95Sa3ptfsogZdwbdNTqDSjfVjdqNUATPNuya4tXq81a4c6ZQqWR9tvtZR1CTZkOdeZYBfeMzGXurfukTob7PGQY2BaSKRLqHbHE76armft1T7EMHieHocSBQjY5n6JG08rMd9WbWi8by5bIfdRbTlZ3HbmM4PNaNFhPA+brrXhdT47LkZvE+BPdox9IA

**Comment by dario-piotrowicz** [comment]
@yjaaidi happy that you agree with that, I hope you didn't spend too much time already on the types changes

The suggestion to create a new issue was based on the fact that it the change would be more than just improving the typing but it's an actual api change, so I was thinking that it could be worth it to open a specific issue for that, especially to formalize the issue and also get suggestions (especially from the angular team), I think it would be a bit clearer instead of continuing the conversation here (but here may be fine too I'd imagine)

Your changes do look great to me :D
(but I would definitely wait for someone else who knows this functionality well to check them, as @JoostK or @AndrewKushnir)

**Comment by dario-piotrowicz** [comment]
Another thing, it may be a bit of a silly one, but if we are going to replace the flags with an object could we maybe rename `self` to something like `onlySelf`, `untilSelf` or similar? I think `self` is a bit less clear, or is that just me? :stuck_out_tongue:

**Comment by AndrewKushnir** [comment]
Quick update: we've discussed this topic with the team and we want to move forward with a PR that would:
- add the support for an object as the second argument in the `inject` and `Injector.get` functions (the `Injector.get` should be a separate commit), while keeping the support for the current format as well for backwards compatibility
- add the deprecation annotation to the type signatures where an old format is used, with a recommendation to use a new object-based format

One more thing thing to consider is the difference between the `inject` and `Injector.get` signatures: the `Injector.get` has an extra argument to provide a default value if none was found in the DI. Ideally, it'd be great to fully align `inject` and `Injector.get`, but it would be a separate effort.

@yjaaidi please let us know if you might be interested in creating a PR to add support for the options object.

**Comment by AndrewKushnir** [comment]
> Another thing, it may be a bit of a silly one, but if we are going to replace the flags with an object could we maybe rename self to something like onlySelf, untilSelf or similar? I think self is a bit less clear, or is that just me?

@dario-piotrowicz good question. I think it'd be valuable to keep the mapping between the current flags (which also map to the decorators like `@Self`, `@Host`, etc) and the inject flags, so it's easier for developers to transition to the options object. However, we should also consider this improvement (may be we can add extra flags, which would be just aliases) and we can have an additional discussion to decide on the final set of flags once the draft PR is ready.

**Comment by yjaaidi** [comment]
Thank you so much for taking care of this @alxhub and sorry again for not responding faster with a PR
